### DISCUSSION TITLE
Feature Request: Method Level Customization

### DISCUSSION THREAD

**Discussion by JulianBirch** [AutoFixture/AutoFixture#1295] [Ideas]
It would be nice to support the ability to customize the fixture at a `MethodInfo` level. The particular example I'm thinking of is adding the ability to inject a factory method in that generates an `IOptions<T>` with specific data. At the moment this is achievable by overriding xUnit's `IEnumerable<object[]> GetData()` and accessing the (deprecated) `Fixture` property, but this isn't ideal.

A minimal solution would be to support the fixture create method taking the target `MethodInfo`. A more complete version would be to support a `MethodCustomizeAttribute` to go with the existing `ParameterCustomizeAttribute`.

**Comment by aivascu** [comment]
@JulianBirch could you provide a code sample of the API you're proposing or maybe an example of how you're currently solving this issue. For now it is not very clear what you're trying to achieve. It sounds a lot like something a mocking framework would be responsible for.

**Comment by Ergamon** [comment]
I guess he is talking about adding customizations at method level.

I e.g. replaced the whole AutoDataAttribute with my own implementation.

Basically it is a copy of the original source looking for customizations starting at method level, going up to class and assembly level. This way I can quickly turn on and off some customizations.

If that is what the original poster wants, I can post it here

**Comment by JulianBirch** [comment]
Sorry, yes, replacing the whole of AutoDataAttribute would also do the trick (I've written my own version that inherits from it, but it needs to access `Fixture`.). Basically, this wouldn't do anything you couldn't do with the regular `CustomizeAttribute` because you can always see the `MethodInfo` from that, but it doesn't make a massive amount of sense for certain things. 

Example 1:

```cs
[AutoData]
[ConsoleLogging]
public Task TestComplexSystem(BigComplex systemUnderTest)
````

Here we have an attribute that causes the test to log to the console. (Admittedly not much use unless you're using NUnit.) We can remove it and just use stub loggers normally, but if we're not sure what's going on on the inside, we can flip logging on  for the whole thing by adding the attribute. 

Example 2
```cs
[AutoData]
[Inject(nameof(CreateDatabaseOptions))]
public Task Test(...)

public static IOptions<DatabaseSettings> CreateDatabaseOptions() 
    => Options.Create(new DatabaseSettings { ConnectionString = "string" });
```

Here we inject in a specific value into anything that needs it. Again, we definitely could do this via a parameter-based attribute, it just feels cleaner to do it this way. (Equally we could avoid the need entirely by freezing a concrete test class that implements
`IOptions<DatabaseSettings>` but I'm not particularly enamoured with that approach either.
